cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2217,6 +2217,73 @@ TEST(Cast, ListToListOptionsPassthru) {
   }
 }
 
+static void CheckStructToStruct(
+    const std::vector<std::shared_ptr<DataType>>& value_types) {
+  for (const auto& src_value_type : value_types) {
+    for (const auto& dest_value_type : value_types) {
+      std::vector<std::string> field_names = {"a"};
+      std::shared_ptr<Array> a1, b1, a2, b2;
+      a1 = ArrayFromJSON(src_value_type, "[1, 2]");
+      a2 = ArrayFromJSON(dest_value_type, "[1, 2]");
+      auto src = StructArray::Make({a1}, field_names).ValueOrDie();
+      auto dest = StructArray::Make({a2}, field_names).ValueOrDie();
+
+      CheckCast(src, dest);
+
+      // Test corner case using children with offsets
+      auto a3 = ArrayFromJSON(src_value_type, "[1, 2, 3]")->Slice(1, 2);
+      auto a4 = ArrayFromJSON(dest_value_type, "[2, 3]");
+      auto slicedSrc = StructArray::Make({a3}, field_names).ValueOrDie();
+      auto slicedDest = StructArray::Make({a4}, field_names).ValueOrDie();
+
+      CheckCast(slicedSrc, slicedDest);
+    }
+  }
+}
+
+TEST(Cast, StructToSameSizedAndNamedStruct) {
+  CheckStructToStruct({int32(), float32(), int64()});
+}
+
+TEST(Cast, StructToSameSizedButDifferentNamedStruct) {
+  std::vector<std::string> field_names = {"a", "b"};
+  std::shared_ptr<Array> a, b;
+  a = ArrayFromJSON(int8(), "[1, 2]");
+  b = ArrayFromJSON(int8(), "[3, 4]");
+  auto src = StructArray::Make({a, b}, field_names).ValueOrDie();
+
+  std::vector<std::string> field_names2 = {"c", "d"};
+  std::shared_ptr<Array> c, d;
+  c = ArrayFromJSON(int8(), "[1, 2]");
+  d = ArrayFromJSON(int8(), "[3, 4]");
+  auto dest = StructArray::Make({c, d}, field_names2).ValueOrDie();
+  auto options = CastOptions{};
+  options.to_type = dest->type();
+
+  ASSERT_RAISES_WITH_MESSAGE(TypeError, "Type error: struct field names do not match",
+                             Cast(src, options));
```

Comment:
CI is failing, use `EXPECT_RAISES_WITH_MESSAGE_THAT(TypeError, ::testing::HasSubstr("..."), ...)` to avoid that